src/execution/periodic_epoch.rs

```diff
@@ -70,6 +71,56 @@ impl PeriodicEpochConfig {
     }
 }
 
+// Struct used to make it clear what we are doing with tracing spans,
+// for internal use of this module.
+struct TracingHelper {
+    step_span: Option<EnteredSpan>,
+    epoch_span: Option<EnteredSpan>,
+    end_of_epoch: bool,
+}
+
+impl TracingHelper {
+    pub fn new() -> Self {
+        Self {
+            step_span: None,
+            epoch_span: None,
+            end_of_epoch: true,
+        }
+    }
+
+    pub fn mark_epoch_end(&mut self) {
+        self.end_of_epoch = true;
+    }
+
+    pub fn end_epoch(&mut self) {
+        if let Some(span) = self.epoch_span.take() {
+            span.exit();
+        }
+    }
+
+    pub fn enter_step(&mut self) {
+        if let Some(parent) = self.epoch_span.as_ref() {
+            self.step_span = Some(tracing::trace_span!(parent: parent, "Step").entered());
+        }
+    }
+
+    pub fn exit_step(&mut self) {
+        if let Some(span) = self.step_span.take() {
+            span.exit();
+        }
+    }
+
+    pub fn update_epoch(&mut self, epoch: &u64) {
+        if self.end_of_epoch {
+            if let Some(span) = self.epoch_span.take() {
+                span.exit();
+            }
+            self.epoch_span = Some(tracing::trace_span!("Epoch", epoch = epoch).entered());
```

Comment:
Yes, should probably find a better name for this, I'm open to suggestions.

The idea of dividing the traces into epochs is to avoid having one giant span for each dataflow, but also avoid having one little span for each `step`, so I used the periodic epoch, but I agree we should not leak this information to the user.

Reply:
Yes, the epoch subdivision was a way to avoid having a single infinite span for each dataflow. The problem being that if we never close the span, the child spans are never correctly grouped.
But I agree that tying the spans to epochs might make things confusing, or as you say give wrong informations.
I found a better way of doing this thing, unrelated to epochs, but with a similar behaviour: open a new span at the beginning of the dataflow run, and close it/open a new one after a fixed (configurable) amount of time.